### PR TITLE
Exposing stream to read object

### PR DESCRIPTION
**Pull request by nhibberd**

### CODE REVIEW

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 1]
```diff
@@ -131,10 +132,14 @@ getObject' =
...
-  let format y = T.concat . TL.toChunks . TL.decodeUtf8 $ y
-  z <- liftIO . sequence $ (\r -> runResourceT (r ^. gorsBody . to bodyResponse $$+- sinkLbs)) <$> resp
-  pure (format <$> z)
+  r <- read' a
+  z <- liftIO . sequence $ (runResourceT . ($$+- sinkLbs)) <$> r
+  pure $ fmap (T.concat . TL.toChunks . TL.decodeUtf8) z
+
+read' :: Address -> AWS (Maybe (ResumableSource (ResourceT IO) BS.ByteString))
+read' a = do
+  r <- getObject' a
```
Comment (charleso) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Should we expose the raw `getObject'` as well?

**Comment by charleso** [comment]
:+1:

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 2]
```diff
@@ -131,10 +132,14 @@ getObject' =
...
 read a = do
-  resp <- getObject' a
-  let format y = T.concat . TL.toChunks . TL.decodeUtf8 $ y
-  z <- liftIO . sequence $ (\r -> runResourceT (r ^. gorsBody . to bodyResponse $$+- sinkLbs)) <$> resp
-  pure (format <$> z)
+  r <- read' a
+  z <- liftIO . sequence $ (runResourceT . ($$+- sinkLbs)) <$> r
+  pure $ fmap (T.concat . TL.toChunks . TL.decodeUtf8) z
+
+read' :: Address -> AWS (Maybe (ResumableSource (ResourceT IO) BS.ByteString))
```
Comment (charleso) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Test might be good. :)

**Comment by domdere** [comment]
nice so we can have a proper streaming `cat` command for `s3` now? 
:+1:

File: mismi-s3/src/Mismi/S3/Commands.hs (hunk 2 above)
Comment (nhibberd) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Would have to rewrite read in the test, which would be the same as testing read which is already tested [here](https://github.com/ambiata/mismi/blob/master/mismi-s3/test/Test/IO/Mismi/S3/Commands.hs)

File: mismi-s3/src/Mismi/S3/Commands.hs (hunk 1 above)
Comment (nhibberd) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Not sure if its needed :grey_question:

**Comment by nhibberd** [comment]
Yeah I think so, cli needs a bit of love in general. Should be done as separate PR

File: mismi-s3/src/Mismi/S3/Commands.hs (hunk 2 above)
Comment (EduardSergeev) [inline mismi-s3/src/Mismi/S3/Commands.hs]: I wonder what would happen if a test for new `read'` never consumes the result or consumes it only partially. It is probably all handled by `conduit` but still would be interesting to see how it works with AWS...

**Pull request merged**